File: C++/Pow(x, n).cpp
/*
Implement pow(x, n), which calculates x raised to the power n (xn).

Example 1:

Input: 2.00000, 10
Output: 1024.00000
*/
class Solution {
public:
    double myPow(double x, int n) {
        long long N = n;
        if (N < 0) {
            x = 1 / x;
            N = -N;
        }
        double ans = 1;
        double current_product = x;
        for (long long i = N; i>0 ; i /= 2) 
        {
            if ((i & 1) == 1)  /* Odd number */ 
            {
                ans = ans * current_product;
            }
            current_product = current_product * current_product;
        }
        return ans;
    }
};

File: Fb_ph/Practice/Python/Pow(x,n).py
'''
Constraints:

-100.0 < x < 100.0
-231 <= n <= 231-1
-104 <= xn <= 104

'''

'''
Wording:-
In multpilication , there is a repetitive word 
so for say 2^10 = 2^5.2^5
for x^n = x^n/2.x^n/2
To avoid duplicating work, we can keep dividing it by 2 and from above we can keep multiplying the number by itself
Question arises for say odd integer, and modification would be required to check if the power is odd or not
One way is recursion but the space complexity is O(n)
However we can still use bit operation as they are faster
'''

'''
TC:- O(Log2(N))
SC:- O(1)
'''
'''
Implement pow(x,n) which calculates x raised to the power n (i.e., xn)
'''
class Solution:
    def myPow(self, x, n):
        m = abs(n)
        ans = 1.0
        while m:
            if m & 1:
                ans *= x
            x *= x
            m >>= 1
        return ans if n >= 0 else 1 / ans
